### PR TITLE
Use Version objects all the way up

### PR DESCRIPTION
**Pull request by samuelhamard**
Issue: #47 

### Successful PR Checklist:
- [x] Tests
- [ ] Documentation (optionally: [run spell checking](https://github.com/peopledoc/septentrion/blob/master/CONTRIBUTING.rst#build-the-documentation))
- [x] Had a good time contributing? (feel free to give some feedback)

### CODE REVIEW

File: septentrion/versions.py [hunk 1]
```diff
@@ -22,3 +22,9 @@ def __eq__(self, other: object) -> bool:
     def __repr__(self) -> str:
         version_string = ".".join(str(n) for n in self._version)
         return f'{self.__class__.__name__}("{version_string}")'
+
+    def __hash__(self) -> int:
+        return hash(self._version)
+
+    def normalized_repr(self) -> str:
+        return ".".join(str(i) for i in self._version)
```
Comment (ewjoachim) [inline septentrion/versions.py]: Does this mean we cannot use "19.04" as version number ?

I think I'd be more comfortable if the version number original string was kept as an attribute of the object ?

File: septentrion/cli.py [hunk 2]
```diff
@@ -229,4 +230,4 @@ def fake(settings: configuration.Settings, version: str):
     all migrations up until the given version (included). This is useful
     when installing septentrion on an existing DB.
     """
-    migrate.create_fake_entries(settings=settings, version=version)
+    migrate.create_fake_entries(settings=settings, version=versions.Version(version))
```
Comment (ewjoachim) [inline septentrion/cli.py]: Shouldn't the `validate_version` callback return Version objects instead ?

File: septentrion/versions.py (hunk 1 above)
Comment (ewjoachim) [reply septentrion/versions.py]: Ok, the more I see it, the more a dataclass makes sense.

Even if it means we'll need a dedicated constructor `Version.from_string()`

**Comment by ewjoachim** [comment]
Looks like we're going in the right direction! I like this. Ping me when you need a review :)

File: septentrion/cli.py [hunk 3]
```diff
@@ -67,9 +67,11 @@ def load_config(ctx: click.Context, param: click.Parameter, value: TextIO) -> No
 
 
 def validate_version(ctx: click.Context, param: Any, value: str):
-    if value and not utils.is_version(value):
+    if value is None:
+        return None
+    if not utils.is_version(value):
```
Comment (moumoutte) [inline septentrion/cli.py]:
```
if utils.is_version(value) is False:
```

File: setup.cfg [hunk 4]
```diff
@@ -27,6 +27,7 @@ install_requires =
     Click
     sqlparse
     colorama
```
Comment (ewjoachim) [inline setup.cfg]: You can make this requirement apply only for py<=3.6
https://www.python.org/dev/peps/pep-0508/

File: setup.cfg [hunk 5]
```diff
@@ -27,6 +27,7 @@ install_requires =
...
+    dataclasses
```
Comment (ewjoachim) [inline setup.cfg]: You can make this requirement apply only for py<=3.6
https://www.python.org/dev/peps/pep-0508/

File: septentrion/cli.py [hunk 6]
```diff
@@ -66,9 +67,11 @@ def load_config(ctx: click.Context, param: click.Parameter, value: TextIO) -> No
...
```
Comment (ewjoachim) [inline septentrion/cli.py]: let's instantiate directly and catch the exception

File: septentrion/db.py [hunk 7]
```diff
@@ -113,23 +113,29 @@ def __call__(self):
...
 
 def get_applied_migrations(
-    settings: configuration.Settings, version: str
+    settings: configuration.Settings, version: versions.Version
 ) -> Iterable[str]:
     with Query(
-        settings=settings, query=query_get_applied_migrations, args=(version,)
+        settings=settings,
+        query=query_get_applied_migrations,
+        args=(version.normalized_repr(),),
```
Comment (ewjoachim) [inline septentrion/db.py]: original_string

File: septentrion/db.py [hunk 8]
```diff
@@ -146,10 +152,12 @@ def create_table(settings: configuration.Settings) -> None:
...
 
-def write_migration(settings: configuration.Settings, version: str, name: str) -> None:
+def write_migration(
+    settings: configuration.Settings, version: versions.Version, name: str
+) -> None:
     Query(
         settings=settings,
         query=query_write_migration,
-        args=(version, name),
+        args=(version.normalized_repr(), name),
```
Comment (ewjoachim) [inline septentrion/db.py]: original_string

File: septentrion/files.py [hunk 9]
```diff
@@ -76,7 +93,7 @@ def get_migrations_files_mapping(
     """
     ignore_symlinks = settings.IGNORE_SYMLINKS
 
-    version_root = settings.MIGRATIONS_ROOT / version
+    version_root = get_version_migration_dir(settings.MIGRATIONS_ROOT, version)
```
Comment (ewjoachim) [inline septentrion/files.py]: .original_name

File: septentrion/utils.py [hunk 10]
```diff
@@ -14,7 +14,7 @@ def sort_versions(iterable: Iterable[str]) -> Iterable[str]:
     >>> sort_versions(["1.0.1", "2.0", "1.0.3"])
     ["1.0.1", "1.0.3", "2.0"]
     """
-    return sorted(iterable, key=versions.Version)
+    return sorted(iterable, key=versions.Version.from_string)
```
Comment (ewjoachim) [inline septentrion/utils.py]: unused

File: septentrion/utils.py [hunk 11]
```diff
@@ -23,7 +23,7 @@ def get_max_version(iterable: Iterable[str]) -> str:
     >>> get_max_version(["1.3", "1.2"])
     "1.3"
     """
-    return max(iterable, key=versions.Version)
+    return max(iterable, key=versions.Version.from_string)
```
Comment (ewjoachim) [inline septentrion/utils.py]: unused

File: septentrion/versions.py [hunk 12]
```diff
@@ -1,24 +1,22 @@
...
-    def __eq__(self, other: object) -> bool:
-        if isinstance(other, Version):
-            return self._version == other._version
-        raise NotImplementedError
-
-    def __repr__(self) -> str:
-        version_string = ".".join(str(n) for n in self._version)
-        return f'{self.__class__.__name__}("{version_string}")'
+    def normalized_repr(self) -> str:
+        return ".".join(str(i) for i in self.version_tuple)
```
Comment (ewjoachim) [inline septentrion/versions.py]: unused

File: septentrion/versions.py [hunk 13]
```diff
@@ -1,24 +1,22 @@
...
+
 from septentrion import exceptions
 
 
+@dataclasses.dataclass(order=True, unsafe_hash=True, frozen=True)
 class Version:
-    def __init__(self, version_string: str):
+
+    version_tuple: tuple = dataclasses.field(compare=True)
+    original_string: str = dataclasses.field(compare=False)
```
Comment (ewjoachim) [inline septentrion/versions.py]: both could be comparable

File: tests/acceptance/test_cli.py [hunk 14]
```diff
@@ -41,4 +41,7 @@ def test_current_database_state(cli_runner, db):
     )
     assert result.exit_code == 0
     assert db_module.is_schema_initialized(settings=settings)
-    assert db_module.get_current_schema_version(settings=settings) == "1.1"
+    assert (
+        db_module.get_current_schema_version(settings=settings).normalized_repr()
```
Comment (ewjoachim) [inline tests/acceptance/test_cli.py]: original_string

File: tests/unit/test_version.py [hunk 15]
```diff
@@ -4,33 +4,48 @@
...
-    assert repr(versions.Version("1.2.3")) == 'Version("1.2.3")'
+    assert (
+        repr(versions.Version.from_string("1.2.3"))
+        == "Version(version_tuple=(1, 2, 3), original_string='1.2.3')"
+    )
+
+
+@pytest.mark.parametrize("received,expected", (("1.2.3", "1.2.3"), ("0.01.2", "0.1.2")))
+def test_version_normalized_repr(received, expected):
+    assert versions.Version.from_string(received).normalized_repr() == expected
```
Comment (ewjoachim) [inline tests/unit/test_version.py]: unused

File: septentrion/versions.py (hunk 13 above)
Comment (samuelhamard) [reply septentrion/versions.py]: For this point I'm not sure. It could be pertinent but i prefer not do it now.

**Pull request merged**